homework/ba-nguyen/lesson-8/javascript-4.md

```diff
@@ -0,0 +1,273 @@
+**Question 1:** The standard JavaScript environment provides another data structure called Set. Like an instance of Map, a set holds a collection of values. Unlike Map, it does not associate other values with those—it just tracks which values are part of the set. A value can be part of a set only once—adding it again doesn’t have any effect.
+Write a class called **_Group_** (since Set is already taken). Like Set, it has add, delete, and has methods. Its constructor creates an empty group, add adds a value to the group (but only if it isn’t already a member), delete removes its argument from the group (if it was a member), and has returns a Boolean value indicating whether its argument is a member of the group.
+Use the === operator, or something equivalent such as indexOf, to determine whether two values are the same.
+Give the class a static from method that takes an iterable object as argument and creates a group that contains all the values produced by iterating over it.
+
+```ts
+class Group<T> {
+  private array: T[];
+  constructor(array: T[]) {
+    this.array = [];
+
+    for (const numb of array) {
+      this.add(numb);
+    }
+  }
+
+  static from<T>(array: T[]): Group<T> {
+    return new Group(array);
+  }
+
+  has(numb: T): boolean {
+    return this.array.indexOf(numb) > -1;
+  }
+
+  add(numb: T): void {
+    if (this.has(numb)) return;
+    this.array.push(numb);
+  }
+
+  delete(numb: T): void {
+    if (this.has(numb)) {
+      const index = this.array.indexOf(numb);
+      this.array.splice(index, 1);
+    }
+  }
+}
+
+let group = Group.from([10, 20]);
+console.log(group.has(10));
+// → true
+console.log(group.has(30));
+// → false
+group.add(10);
+group.delete(10);
+console.log(group.has(10));
+// → false
+```
+
+**Question 2:** Write an expression that matches only JavaScript-style numbers. It must support an optional minus or plus sign in front of the number, the decimal dot, and exponent notation—5e-3 or 1E10—again with an optional sign in front of the exponent. Also note that it is not necessary for there to be digits in front of or after the dot, but the number cannot be a dot alone. That is, .5 and 5. are valid JavaScript numbers, but a lone dot isn’t.
+
+```js
+// Fill in this regular expression.
+let number =
+  /^(\+|-)*[\d]+((e|E)(\+|-)*[\d]+)*(\.(\+|-)*[\d]+(e|E)*((\+|-)*[\d]+)*){0,1}$/;
```

Comment:
![Screen Shot 2022-08-23 at 14 14 37](https://user-images.githubusercontent.com/29374426/186095085-58cd6e5d-c90b-4568-866e-d4f302f2c61d.png)
not pass some test case